lib/Sema/ConstraintSystem.cpp

```diff
@@ -1564,12 +1564,11 @@ Type simplifyTypeImpl(ConstraintSystem &cs, Type type, Fn getFixedTypeFn) {
       if (newBase.getPointer() == depMemTy->getBase().getPointer())
         return type;
 
-      Type lookupBaseType = newBase->getLValueOrInOutObjectType();
-
       // Dependent member types should only be created for associated types.
       auto assocType = depMemTy->getAssocType();
-      assert(depMemTy->getAssocType());
+      assert(depMemTy->getAssocType() && "Expected associated type!");
 
+      Type lookupBaseType = newBase->getLValueOrInOutObjectType();
```

Comment:
I think it's weird that we can end up getting a member type of an inout or lvalue type, also. I remember thinking this was dead code but it isn't.